examples/generative/cyclegan.py

```diff
@@ -0,0 +1,695 @@
+            ax[i, 1].imshow(prediction)
+            ax[i, 0].set_title("Input image")
+            ax[i, 1].set_title("Translated image")
+            ax[i, 0].axis("off")
+            ax[i, 1].axis("off")
+
+            prediction = keras.preprocessing.image.array_to_img(prediction)
+            prediction.save(
+                "generated_img_{i}_{epoch}.png".format(i=i, epoch=epoch + 1)
+            )
+        plt.show()
+        plt.close()
+
+
+"""
+## Create a callback that periodically saves the models and optimizers
+"""
+
+
+class ModelCheckpointer(keras.callbacks.Callback):
+    """Saves model weights and optimizers"""
+
+    def __init__(self, checkpoint_path, save_freq=5):
+        self.checkpoint_path = checkpoint_path
+        self.save_freq = save_freq
+
+    def on_epoch_end(self, epoch, logs=None):
+        if epoch == 0:
+            ckpt = tf.train.Checkpoint(
+                generator_g=self.model.gen_G,
+                generator_f=self.model.gen_F,
+                discriminator_x=self.model.disc_X,
+                discriminator_y=self.model.disc_Y,
+                generator_g_optimizer=self.model.gen_G_optimizer,
+                generator_f_optimizer=self.model.gen_F_optimizer,
+                discriminator_x_optimizer=self.model.disc_X_optimizer,
+                discriminator_y_optimizer=self.model.disc_Y_optimizer,
+            )
+            self.ckpt_manager = tf.train.CheckpointManager(
+                ckpt, self.checkpoint_path, max_to_keep=5
+            )
+
+            if self.ckpt_manager.latest_checkpoint:
+                ckpt.restore(self.ckpt_manager.latest_checkpoint)
+                print("Latest checkpoint restored successfully!")
+        ckpt_save_path = self.ckpt_manager.save()
+        print(
+            "\nSaving checkpoint for epoch {} at {}".format(epoch + 1, ckpt_save_path)
+        )
+
+
+"""
+## Train the end-to-end model
+"""
+
+
+# Loss function for evaluating adverserial loss
+adv_loss_fn = keras.losses.MeanSquaredError()
+
+# Define the loss function for the generators
+def generator_loss_fn(fake):
+    fake_loss = adv_loss_fn(tf.ones_like(fake), fake)
+    return fake_loss
+
+
+# Define the loss function for the discriminators
+def discriminator_loss_fn(real, fake):
+    real_loss = adv_loss_fn(tf.ones_like(real), real)
+    fake_loss = adv_loss_fn(tf.zeros_like(fake), fake)
+    return (real_loss + fake_loss) * 0.5
+
+
+# Optimizers with parameters values as recommended
+gen_G_optimizer = keras.optimizers.Adam(learning_rate=2e-4, beta_1=0.5)
+gen_F_optimizer = keras.optimizers.Adam(learning_rate=2e-4, beta_1=0.5)
+disc_X_optimizer = keras.optimizers.Adam(learning_rate=2e-4, beta_1=0.5)
+disc_Y_optimizer = keras.optimizers.Adam(learning_rate=2e-4, beta_1=0.5)
+
+
+# Create cycle gan model
+cycle_gan_model = CycleGan(
+    generator_G=gen_G, generator_F=gen_F, discriminator_X=disc_X, discriminator_Y=disc_Y
+)
+
+# Compile the model
+cycle_gan_model.compile(
```

Comment:
You'd save space by inlining the optimizer definitions as well as `adv_loss_fn` in this call